gio/src/extension.rs

```diff
@@ -0,0 +1,200 @@
+
+impl fmt::Display for IOExtensionPoint {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "IOExtensionPoint")
+    }
+}
+
+impl FromGlibPtrNone<*mut ffi::GIOExtensionPoint> for IOExtensionPoint {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *mut ffi::GIOExtensionPoint) -> Self {
+        assert!(!ptr.is_null());
+        IOExtensionPoint(ptr::NonNull::new_unchecked(ptr))
+    }
+}
+
+impl<'a> ToGlibPtr<'a, *mut ffi::GIOExtensionPoint> for &'a IOExtensionPoint {
+    type Storage = &'a IOExtensionPoint;
+
+    #[inline]
+    fn to_glib_none(&self) -> Stash<'a, *mut ffi::GIOExtensionPoint, &'a IOExtensionPoint> {
+        Stash(self.0.as_ptr() as *mut ffi::GIOExtensionPoint, *self)
+    }
+}
+
+impl IOExtensionPoint {
+    #[doc(alias = "g_io_extension_point_register")]
+    pub fn register(name: &str) -> Self {
+        unsafe { from_glib_none(ffi::g_io_extension_point_register(name.to_glib_none().0)) }
+    }
+
+    #[doc(alias = "g_io_extension_point_lookup")]
+    pub fn lookup(name: &str) -> Option<Self> {
+        unsafe {
+            let ep = ffi::g_io_extension_point_lookup(name.to_glib_none().0);
+            if ep.is_null() {
+                None
+            } else {
+                Some(from_glib_none(ep))
+            }
+        }
+    }
+
+    #[doc(alias = "g_io_extension_point_get_extensions")]
+    pub fn extensions(&self) -> Vec<IOExtension> {
+        let mut res = Vec::new();
+        unsafe {
+            let mut l = ffi::g_io_extension_point_get_extensions(self.0.as_ptr());
+            while !l.is_null() {
+                let e: *mut ffi::GIOExtension = Ptr::from((*l).data);
+                res.push(from_glib_none(e));
+                l = (*l).next;
+            }
+        }
+        res
+    }
+
+    #[doc(alias = "g_io_extension_point_get_extension_by_name")]
+    pub fn extension_by_name(&self, name: &str) -> Option<IOExtension> {
+        unsafe {
+            let e = ffi::g_io_extension_point_get_extension_by_name(
+                self.0.as_ptr(),
+                name.to_glib_none().0,
+            );
+            if e.is_null() {
+                None
+            } else {
+                Some(from_glib_none(e))
+            }
+        }
+    }
+
+    #[doc(alias = "g_io_extension_point_get_required_type")]
+    pub fn required_type(&self) -> Type {
+        unsafe { from_glib(ffi::g_io_extension_point_get_required_type(self.0.as_ptr())) }
+    }
+
+    #[doc(alias = "g_io_extension_point_set_required_type")]
+    pub fn set_required_type(&mut self, type_: Type) {
+        unsafe { ffi::g_io_extension_point_set_required_type(self.0.as_ptr(), type_.into_glib()) }
+    }
+
+    #[doc(alias = "g_io_extension_point_implement")]
+    pub fn implement(
+        extension_point_name: &str,
+        type_: Type,
+        extension_name: &str,
+        priority: i32,
+    ) -> IOExtension {
+        unsafe {
+            let e = ffi::g_io_extension_point_implement(
```

Comment:
This should probably assert that `type_` is a / implements the required type and that the extension point actually exists

Reply:
This is done internally by the glib function, do we need to duplicate it? Or do you mean that we have a way to enforce it with the type system?

Reply:
Well, what will happen if you call the C function wrong here? `g_critical()` and then it panics because the returned pointer is `NULL`?

Reply:
`g_warning` and then panics because of `NULL`... shall I make it fallible and return `Option`?

Reply:
Updated to do that

Reply:
No, it should panic in one way or another. It's a programmer error, not something the programmer is supposed to handle at runtime.

Reply:
mmm, but the glib side are warnings... I am wondering if that is intentional so that a broken extension does not crash the entire app

Reply:
Just like `g_return_val_if_fail()`. If that ever happens then your application is not going to behave as it should.

Reply:
Note also that returning an `Option` does not make the code harder for the caller, since normally you do not really need to check that return value

Reply:
Yes and then the error is completely silently ignored :)

Reply:
`g_return_val_if_fail` is a critical you can turn into an assert, these ones are more legitimate runtime warnings IMHO.

If eg my app has an extension point and at some point I bump the ABI and require extensions to implement a new GType, I want my app to not crash even if there are stray .so in the folder and I want my app to log that the extension is ignored.

Sure, on the caller side the error will be ignored, but then again the C api returns a pointer just for the far fetched case you want to call it multiple times.

Reply:
Seems like a case where you clearly want your application to stop functioning because the whole environment is not set up the way it's expected to, and it won't work as expected (i.e. the extension implementor wouldn't be used).

Reply:
I do not agree: if eg a third party nautilus plugin that adds some simple convenience as an extra item in the popup menu is outdated with a new version of nautilus, I do not want the entire nautilus to crash: I want the plugin ignored and if I really care about that plugin, I can see in the logs that it has been ignored.

Also, the policy here is set in gio, which logs a warning and ignores the extension, I do not think the Rust bindings should enforce a different policy in this regard.

With that said, I promise this is my last comment on the matter :-)
If you still think we should panic, I will add back the assert and push an updated patch.

Reply:
Ok, let's keep it like this for now then.